### PR TITLE
Use host kernel with Nvidia Orin guest

### PR DESCRIPTION
**Pull request by juliuskoskela**
## Description of changes

Continuation of[ PR](https://github.com/tiiuae/ghaf/pull/394). We implement virtual machine guest context for Nvidia Orin.

- [x] Can build Nvidia Orin host's kernel (Linux 5.1 + Tegra patches) for the guest.

## Checklist for things done

- [x] Summary of the proposed changes in the PR description
- [x] More detailed description in the commit message(s)
- [ ] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [ ] Test procedure described (or includes tests). Select one or more:
  - [ ] Tested on Lenovo X1 `x86_64`
  - [x] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [x] Author has run `nix flake check --accept-flake-config` and it passes
- [x] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status

## Testing

1. Define a VM in a Ghaf downstream flake using microvm Nix.
2. In the guest vm's configuration enable `ghaf.hardware.nvidia.virtualizatin.guest.enable = true`
3. After rebuilding the host, run the vm using `micorvm -r <vm-name>.
4. Inside the VM check that Linux version is 5.1

### CODE REVIEW

**Review by avnik — CHANGES_REQUESTED** [review]
All other LGTM.

@vilvo are this change OK from security perspective? (using same host/guest kernels)

**Comment by vilvo** [comment]
> @vilvo are this change OK from security perspective? (using same host/guest kernels)

Security-wise this structure will require refactoring to adapt hardened kernel and hardened kernel configuration in guest via fragment instead of `extraStructuredConfig`. See #439 
We (me and @vunnyso) could not make minimal kernel with own baseline kernel config work with `extraStructuredConfig` because of nixpkgs kernel builder infrastructure limitations.

**Comment by juliuskoskela** [comment]
> > @vilvo are this change OK from security perspective? (using same host/guest kernels)
> 
> Security-wise this structure will require refactoring to adapt hardened kernel and hardened kernel configuration in guest via fragment instead of `extraStructuredConfig`. See #439 We (me and @vunnyso) could not make minimal kernel with own baseline kernel config work with `extraStructuredConfig` because of nixpkgs kernel builder infrastructure limitations.

I would rather suggest we stay away from our own solutions and prefer Nix tooling or upstreaming to Nix tooling. Did we actually get to the bottom of this issue? If I investigate how the kernel is packaged in Jetpack-nixos I observe the following:

https://github.com/anduril/jetpack-nixos/blob/master/kernel/default.nix

The above file packages the kernel using `pkgs.buildLinux`. Inside the function patches, extraStructuredConfigs etc. are passed (mind the use of `lib.kernel` when defining with extraStructuredConfig).

https://github.com/anduril/jetpack-nixos/blob/master/default.nix

The above file exposes the kernel [here](https://github.com/anduril/jetpack-nixos/blob/0cf783ad8be0b13a65f0debeeedf9263c4609da0/default.nix#L91) (note the use of `pkgs.linuxPackagesFor` for applying overlays).

Even after the above packaging I can still add kernel configurations and patches successfully.

File: targets/nvidia-jetson-orin/default.nix [hunk 1]
```diff
@@ -38,7 +38,7 @@
     ];
     hostConfiguration = lib.nixosSystem {
       inherit system;
-      specialArgs = {inherit lib;};
+      specialArgs = {inherit lib jetpack-nixos;};
```
[the remark this answers is not in the document]
Comment (juliuskoskela) [reply targets/nvidia-jetson-orin/default.nix:41]: Yes I agree that's a better idea. We will need a variety of things straight from the inputs inside the modules anyways. It makes sense to pass the inputs as a whole from this point.

**Comment by juliuskoskela** [comment]
I think we don't necessarily need to hurry merging this. Important decisions concerning how we define our kernels is ongoing for example in PR #439 which could possibly affect how this feature is implemented. Furthermore this is all scaffolding for GPU passthrough on Nvidia Jetson which is ongoing and will take a while. By then we might have moved to a newer kernel and this module will have become either obsolete or we will attach a different kernel configuration to it.

**Pull request closed without merging**